devices/__init__.py

```diff
@@ -19,7 +19,141 @@
+        this in a separate method from `__init__` has the advantage that it can be used to reset those
+        variables at a later stage, without having to write the same code again.
+
+        Following this scheme, inheriting from `StateMachineDevice` also provides the possibility
+        for users of the class to override the states, the transitions, the initial state and
+        even the data. For states, transitions and data, dicts need to be passed to the constructor,
+        for the initial state that should be a string.
+
+        All these overrides can be used to define device setups to describe certain scenarios more easily.
+
+        :param override_states: Dict with one entry per state. Only states defined in the state machine are allowed.
+        :param override_transitions: Dict with (state, state) tuples as keys and callables as values.
+        :param override_initial_state: The initial state.
+        :param override_initial_data: A dict that contains data members that should be overwritten on construction.
+        """
+        super(StateMachineDevice, self).__init__()
+
+        self._initialize_data()
+        self._override_data(override_initial_data)
+
+        self._csm = StateMachine({
+            'initial': self._get_initial_state() if override_initial_state is None else override_initial_state,
+            'states': self._get_final_state_handlers(override_states),
+            'transitions': self._get_final_transition_handlers(override_transitions)
+        }, context=self)
+
+        self.addProcessor(self._csm)
+
+    def _get_state_handlers(self):
+        """
+        Implement this method to return a dict-like object with state handlers (see core.statemachine.State) for each
+        state of the state machine. The default implementation raises a NotImplementedError.
+
+        :return: A dict-like object containing named state handlers.
+        """
+        raise NotImplementedError('_get_state_handlers must be implemented in a StateMachineDevice.')
+
+    def _get_initial_state(self):
+        """
+        Implement this method to return the initial state of the internal state machine. The default implementation
+        raises a NotImplementedError.
+
+        :return: The initial state of the state machine.
+        """
+        raise NotImplementedError('_get_initial_state must be implemented in a StateMachineDevice.')
+
+    def _get_transition_handlers(self):
+        """
+        Implement this method to return transition handlers for the internal state machine. The keys should be
+        (state, state)-tuples and the values functions that return true if the transition should be triggered.
+
+        :return: A dict-like object containing transition handlers.
+        """
+        raise NotImplementedError('_get_transition_handlers must be implemented in a StateMachineDevice.')
+
+    def _initialize_data(self):
```

Comment:
I was going to say that documentation for this would have been useful. But you beat me to it :)